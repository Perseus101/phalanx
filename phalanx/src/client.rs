use std::string::FromUtf8Error;

use async_trait::async_trait;
use err_derive::Error;

use reqwest::{Client as ReqwestClient, Error as ReqwestError, Response};

use crate::util::AsyncTryFrom;

pub struct Client {
    pub client: ReqwestClient,
    pub url: String,
}

impl Client {
    pub fn new(client: ReqwestClient, url: String) -> Self {
        Client { client, url }
    }

    pub fn url(url: &str) -> Self {
        Self::new(ReqwestClient::default(), String::from(url))
    }

    pub fn format_url(&self, relative_url: &str) -> String {
        format!("{}{}", self.url, relative_url)
    }
}

impl From<String> for Client {
    fn from(url: String) -> Self {
        Self::new(ReqwestClient::default(), url)
    }
}

impl<'a> From<&'a str> for Client {
    fn from(url: &'a str) -> Self {
        Self::url(url)
    }
}

pub trait PhalanxClient {
    fn client(&self) -> &Client;
}

pub struct PhalanxResponse(Response);

impl From<Response> for PhalanxResponse {
    fn from(res: Response) -> Self {
        PhalanxResponse(res)
    }
}

#[derive(Debug, Error)]
pub enum PhalanxClientError {
    #[error(display = "error making request")]
    ReqwestError(#[error(source)] ReqwestError),
    #[error(display = "error parsing request")]
    ParseError(Box<dyn std::error::Error>),
}

impl From<FromUtf8Error> for PhalanxClientError {
    fn from(err: FromUtf8Error) -> Self {
        PhalanxClientError::ParseError(err.into())
    }
}
#[async_trait]
impl AsyncTryFrom<PhalanxResponse> for String {
    type Error = PhalanxClientError;

    async fn try_from(res: PhalanxResponse) -> Result<Self, Self::Error> {
        Ok(String::from_utf8(Vec::from(&res.0.bytes().await?[..]))?)
    }
}

#[async_trait]
impl AsyncTryFrom<PhalanxResponse> for () {
    type Error = PhalanxClientError;

    async fn try_from(res: PhalanxResponse) -> Result<Self, Self::Error> {
        res.0.error_for_status()?;
        Ok(())
    }
}
