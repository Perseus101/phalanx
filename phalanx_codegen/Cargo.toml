[package]
name = "phalanx_codegen"
version = "0.1.0"
authors = ["Colin Moore <colin@moore.one>"]
edition = "2018"

[lib]
proc-macro = true

[dependencies]
syn = { version = "1.0.17", features = ["full"] }
proc-macro-error = "0.4.12"
quote = "1.0.3"
proc-macro2 = "1.0.9"
regex = "1.4.3"
lazy_static = "1.4.0"

[dev-dependencies]
phalanx = { path = "../phalanx" }
actix-web = "3.3.2"
trybuild = "1.0.38"
serde = "1.0.119"
