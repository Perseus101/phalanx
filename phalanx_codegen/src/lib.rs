use proc_macro::TokenStream;

use proc_macro_error::proc_macro_error;
use syn::{parse_macro_input, DeriveInput};

mod derive;
mod route;
mod service;

macro_rules! method_macro {
    (
        $($variant:ident, $method:ident,)+
    ) => {
        $(
            #[proc_macro_attribute]
            pub fn $method(_: TokenStream, input: TokenStream) -> TokenStream {
                // These attributes are parsed and used by the phalax_server proc_macro
                input
            }
        )+
    };
}

method_macro! {
    Get,       get,
    Post,      post,
    Put,       put,
    Delete,    delete,
    Head,      head,
    Connect,   connect,
    Options,   options,
    Trace,     trace,
    Patch,     patch,
}

#[proc_macro_attribute]
#[proc_macro_error]
pub fn phalanx(attr: TokenStream, input: TokenStream) -> TokenStream {
    service::Service::from_tokens(attr, input)
}

#[proc_macro_derive(PhalanxClient, attributes(client))]
pub fn derive_serialize(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as DeriveInput);

    match derive::derive_serialize_inner(input) {
        Err(e) => e.to_compile_error().into(),
        Ok(s) => s.into(),
    }
}
