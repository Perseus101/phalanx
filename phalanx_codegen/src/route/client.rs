use proc_macro2::TokenStream as TokenStream2;
use quote::{quote, ToTokens};

use super::Route;
pub struct ClientRoute(Route);

impl From<Route> for ClientRoute {
    fn from(route: Route) -> Self {
        ClientRoute(route)
    }
}

impl ToTokens for ClientRoute {
    fn to_tokens(&self, tokens: &mut TokenStream2) {
        let args = &self.0.args;
        let fn_name = &self.0.ident;
        let raw_ret_type = &self.0.ret_type;
        let attrs = &self.0.attrs;
        let route = &self.0.route_attr.route;
        let method = self.0.route_attr.method_ident_lower();

        // Get just the arguments which affect the path of the request
        let path_args = super::split_args(&self.0.path_args);

        let format_args: Vec<TokenStream2> = path_args
            .iter()
            .map(|(ident, _)| quote! { #ident = #ident })
            .collect();

        let format_url = if format_args.len() > 0 {
            quote! { &format!( #route , #(#format_args),* ) }
        } else {
            quote! { #route }
        };

        // Ensure the type is a result type
        let ret_type = match raw_ret_type {
            syn::ReturnType::Default => {
                quote! { () }
            }
            syn::ReturnType::Type(_, ty) => {
                quote! { #ty }
            }
        };

        let (content_type, payload) = if let Some(payload) = &self.0.payload_arg {
            match payload.pat.as_ref() {
                syn::Pat::Ident(ident) => (
                    quote! {
                        let __content_type = phalanx::client::ContentType::from(&#ident);
                    },
                    quote! {
                        .header("content-type", __content_type.header_value())
                        .body({
                            let body: phalanx::reexports::Body = std::convert::TryFrom::try_from(#ident)?;
                            body
                        })
                    },
                ),
                pat => panic!("Unknown pattern: {:?}", pat),
            }
        } else {
            (quote! {}, quote! {})
        };

        let stream = quote! {
            #(#attrs)*
            pub async fn #fn_name ( &self, #(#args),* ) -> Result< #ret_type , Box<dyn std::error::Error> > {
                let __client  = phalanx::client::PhalanxClient::client(self);
                #content_type
                let __req = __client.client. #method (&__client.format_url( #format_url )) #payload;
                let __res = phalanx::client::PhalanxResponse::from(__req.send().await?);
                Ok(<#ret_type as phalanx::util::AsyncTryFrom<phalanx::client::PhalanxResponse>>::try_from(__res).await?)
            }
        };

        tokens.extend(stream);
    }
}
