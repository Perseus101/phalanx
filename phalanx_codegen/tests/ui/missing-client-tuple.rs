use phalanx::client::Client as InnerClient;
use phalanx_codegen::PhalanxClient;

#[derive(PhalanxClient)]
struct Client(InnerClient);

fn main() {}
