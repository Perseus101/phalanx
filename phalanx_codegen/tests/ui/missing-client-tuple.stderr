error: Missing a `#[client]` attribute on the client field
 --> $DIR/missing-client-tuple.rs:4:10
  |
4 | #[derive(PhalanxClient)]
  |          ^^^^^^^^^^^^^
  |
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
